### ISSUE TITLE
Invalid Operation @ Player.InventoryOnUpdate

### ISSUE THREAD

**Issue by Lizzaran** [AeonLucid/POGOLib#14]
Gets triggered if you for example recycle an item.

Error Details:

> System.InvalidOperationException was unhandled
  HResult=-2146233079
  Message=Sequence contains no matching element
  Source=System.Core
  StackTrace:
       at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
       at POGOLib.Pokemon.Player.InventoryOnUpdate(Object sender, EventArgs eventArgs) in C:\Users\Lizzaran\Desktop\PoGoBot\POGOLib\Pokemon\Player.cs:line 69
       at POGOLib.Pokemon.Inventory.OnUpdate() in C:\Users\Lizzaran\Desktop\PoGoBot\POGOLib\Pokemon\Inventory.cs:line 44
       at POGOLib.Pokemon.Inventory.set_InventoryItems(RepeatedField`1 value) in C:\Users\Lizzaran\Desktop\PoGoBot\POGOLib\Pokemon\Inventory.cs:line 38
       at POGOLib.Net.RpcClient.HandleDefaultResponses(RepeatedField`1 returns) in C:\Users\Lizzaran\Desktop\PoGoBot\POGOLib\Net\RpcClient.cs:line 440
       at POGOLib.Net.RpcClient.HandleResponseEnvelope(ResponseEnvelope responseEnvelope) in C:\Users\Lizzaran\Desktop\PoGoBot\POGOLib\Net\RpcClient.cs:line 403
       at POGOLib.Net.RpcClient.SendRemoteProcedureCall(Request request) in C:\Users\Lizzaran\Desktop\PoGoBot\POGOLib\Net\RpcClient.cs:line 383
       at PoGoBot.Logic.Automation.Modules.Inventory.Recycle.ItemRecycle(ItemId itemId, Int32 amount) in C:\Users\Lizzaran\Desktop\PoGoBot\PoGoBot.Logic\Automation\Modules\Inventory\Recycle.cs:line 65
       at PoGoBot.Logic.Automation.Modules.Inventory.Recycle.OnThink() in C:\Users\Lizzaran\Desktop\PoGoBot\PoGoBot.Logic\Automation\Modules\Inventory\Recycle.cs:line 44
       at PoGoBot.Logic.Automation.Bot.Think() in C:\Users\Lizzaran\Desktop\PoGoBot\PoGoBot.Logic\Automation\Bot.cs:line 47
       at PoGoBot.Console.Program.Main() in C:\Users\Lizzaran\Desktop\PoGoBot\PoGoBot.Console\Program.cs:line 51
       at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
       at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
       at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
       at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
       at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
       at System.Threading.ThreadHelper.ThreadStart()
  InnerException:

**Comment by MajdajkD** [comment]
It works fine for me with the following code:

 session.RpcClient.SendRemoteProcedureCall(RequestType.GetPlayer);
            var inventoryResponse = session.RpcClient.SendRemoteProcedureCall(RequestType.GetInventory);
            GetInventoryResponse inventory = GetInventoryResponse.Parser.ParseFrom(inventoryResponse);

Dictionary<POGOProtos.Inventory.ItemId, int> items = new Dictionary<POGOProtos.Inventory.ItemId, int>();

                foreach (var inventoryItem in inventory.InventoryDelta.InventoryItems)
                {
                    var item = inventoryItem.InventoryItemData.Item;
                    if (item != null && item.Count > 10 &&
                        (item.ItemId == POGOProtos.Inventory.ItemId.ItemRevive ||
                        item.ItemId == POGOProtos.Inventory.ItemId.ItemPotion ||
                        item.ItemId == POGOProtos.Inventory.ItemId.ItemHyperPotion ||
                        item.ItemId == POGOProtos.Inventory.ItemId.ItemRazzBerry ||
                        item.ItemId == POGOProtos.Inventory.ItemId.ItemSuperPotion ||
                        item.ItemId == POGOProtos.Inventory.ItemId.ItemPokeBall))
                    {
                        items.Add(item.ItemId, item.Count);
                    }
                }

                foreach (var item in items)
                {

                    var recycleResponse = session.RpcClient.SendRemoteProcedureCall(new Request
                    {
                        RequestType = RequestType.RecycleInventoryItem,
                        RequestMessage = new RecycleInventoryItemMessage
                        {
                            ItemId = item.Key,
                            Count = item.Value - 15
                        }.ToByteString()
                    });

                    var responseMessage = RecycleInventoryItemResponse.Parser.ParseFrom(recycleResponse);

                    Console.ForegroundColor = ConsoleColor.DarkMagenta;
                    Log.Info($"Recycled items:{item.Key}, Remaining: {responseMessage.NewCount} ");
                    Console.ForegroundColor = ConsoleColor.White;

                    Thread.Sleep(new TimeSpan(0, 0, 2).Milliseconds);
                }

**Comment by Lizzaran** [comment]
@MajdajkD 

 This has to do with how the inventory updates are handled by the library, mentioned in #15 .
As a temporary fix until I write the new logic for the inventory, a small workaround would be:

```
private void InventoryOnUpdate(object sender, EventArgs eventArgs)
{
    var stats = Inventory?.InventoryItems?.FirstOrDefault()?.InventoryItemData?.PlayerStats;
    if (stats != null)
    {
        Stats = stats;
    }
}
```

**Comment by MajdajkD** [comment]
I believe this has to do with the Async nature of the requests, I've refactored my code and implemented proper lock for the session for various activities (like catch, spin, recycle, walk, inventory updates, etc. ), it solved all my problem. Sometimes I get response code 3 or 0, but I think it happens because I did not lock all transactions properly in the POGOLib project. (And I catch it in empty catch block... THUGLIFE :) )
